### PR TITLE
FFmpeg5 and CBS library

### PR DESCRIPTION
**Pull request by brad-richardson**
## Description
Updates ffmpeg source to track 5.1 release branch. Also moves sunshine's internal CBS library to ensure we reference the same source files. See discussion on LizardByte/Sunshine#509 for more context

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [X] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [X] I want maintainers to keep my branch updated

### CODE REVIEW

**Review by ReenigneArcher — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 # JetBrains project folder
 .idea/
+build/
```
Comment (ReenigneArcher) [inline .gitignore:3]:
```txt
# Distribution / packaging
build/

# JetBrains project folder
.idea/
```

File: ffmpeg_patches/cbs/avconfig.patch [hunk 2]
```diff
@@ -0,0 +1,41 @@
+diff --git a/libavutil/avconfig.h b/libavutil/avconfig.h
```
Comment (ReenigneArcher) [inline ffmpeg_patches/cbs/avconfig.patch:1]: If you have any idea the purpose or original source of these patches, can you add some details to the top like in the other patches?

(Applies to all 4 patch files)

File: ffmpeg_patches/cbs/avconfig.patch [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+new file mode 100644
+index 0000000000..a778bdd9f2
+--- /dev/null
++++ b/libavutil/avconfig.h
+@@ -0,0 +1,34 @@
++// [sunshine] Copied, generated file. Needs to exist in avutil/ and avcodec/
```
Comment (ReenigneArcher) [inline ffmpeg_patches/cbs/avconfig.patch:7]: this line won't be necessary (not sure if patch file needs to be re-created)

File: ffmpeg_patches/cbs/avconfig.patch [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
++// It's a little-endian target architecture
++#define AV_HAVE_BIGENDIAN 0
++
++#else
++// https://manhnt.github.io/programming_technique/2018/08/15/oneline-macro-endian-check.html
++#define AV_HAVE_BIGENDIAN (*(uint16_t *)"\0\xff" < 0x0100)
++#endif
++
++#endif
+\ No newline at end of file
```
Comment (ReenigneArcher) [inline ffmpeg_patches/cbs/avconfig.patch:41]: this looks odd (I've not seen this before)

File: ffmpeg_patches/cbs/config.patch [hunk 5]
```diff
@@ -0,0 +1,82 @@
+diff --git a/libavcodec/config.h b/libavcodec/config.h
+new file mode 100644
+index 0000000000..8c42d1c51c
+--- /dev/null
++++ b/libavcodec/config.h
+@@ -0,0 +1,34 @@
++// [sunshine] Copied, generated file. Needs to exist in avutil/ and avcodec/
```
Comment (ReenigneArcher) [inline ffmpeg_patches/cbs/config.patch:7]: this line won't be necessary (not sure if patch file needs to be re-created)

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,147 @@
+cmake_minimum_required(VERSION 3.0)
```
Comment (ReenigneArcher) [inline CMakeLists.txt:1]: The repo may (possibly) contain additional dependencies besides ffmpeg in the future...

Perhaps we can rename to `ffmpeg_cbs.cmake` or something? I don't know if we should have a directory where we place cmake files.

**Review by ReenigneArcher — COMMENTED** [review]
Here are my initial thoughts.

File: CMakeLists.txt (hunk 6 above)
Comment (brad-richardson) [reply CMakeLists.txt:1]: I'm not sure how to do this... do I just have a dummy top-level CMakeLists.txt for now? cmake won't pick it up otherwise right?

Comment (ReenigneArcher) [reply CMakeLists.txt:1]: I think you can call it directly.

`cmake <options> ffmpeg_cbs.cmake`

Comment (ReenigneArcher) [reply CMakeLists.txt:1]: If that doesn't work...

`cmake -P <other options> ffmpeg_cbs.cmake` I think will work...

Or worst case, in the workflow, rename the file back to `CMakeLists.txt` before executing `cmake`.

Comment (ReenigneArcher) [reply CMakeLists.txt:1]: how about this?

**CMakeLists.txt**
```txt
#if (FFMPEG_CBS)
  include (${CMAKE_CURRENT_SOURCE_DIR}/ffmpeg_cbs.cmake)
#endif()
```

**command**
```bash
cmake -DFFMPEG_CBS=ON ...
```

File: cmake/libcbs.pc.in [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+prefix=@CMAKE_INSTALL_PREFIX@
+exec_prefix="${prefix}"
+libdir="${prefix}/lib"
+includedir="${prefix}/include"
+
+Name: @PROJECT_NAME@
+Description: @CMAKE_PROJECT_DESCRIPTION@
+Version: @PROJECT_VERSION@
+Cflags: -I"${includedir}"
+Libs: -L"${libdir}" -lcbs
```
Comment (brad-richardson) [inline cmake/libcbs.pc.in:12]: This doesn't seem to populate fully but I figure it's good to produce anyway

File: cmake/ffmpeg_cbs.cmake [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+configure_file(${AVCODEC_GENERATED_SRC_PATH}/hevc.h ${CBS_INCLUDE_PATH}/hevc.h COPYONLY)
+configure_file(${AVCODEC_GENERATED_SRC_PATH}/mathops.h ${CBS_INCLUDE_PATH}/mathops.h COPYONLY)
+configure_file(${AVCODEC_GENERATED_SRC_PATH}/packet.h ${CBS_INCLUDE_PATH}/packet.h COPYONLY)
+configure_file(${AVCODEC_GENERATED_SRC_PATH}/sei.h ${CBS_INCLUDE_PATH}/sei.h COPYONLY)
+configure_file(${AVCODEC_GENERATED_SRC_PATH}/vlc.h ${CBS_INCLUDE_PATH}/vlc.h COPYONLY)
+configure_file(${FFMPEG_GENERATED_SRC_PATH}/config.h ${CMAKE_BINARY_DIR}/include/config.h COPYONLY)
+configure_file(${FFMPEG_GENERATED_SRC_PATH}/libavutil/x86/asm.h ${CMAKE_BINARY_DIR}/include/libavutil/x86/asm.h COPYONLY)
+configure_file(${FFMPEG_GENERATED_SRC_PATH}/libavutil/x86/intmath.h ${CMAKE_BINARY_DIR}/include/libavutil/x86/intmath.h COPYONLY)
+configure_file(${FFMPEG_GENERATED_SRC_PATH}/libavutil/arm/intmath.h ${CMAKE_BINARY_DIR}/include/libavutil/arm/intmath.h COPYONLY)
+configure_file(${FFMPEG_GENERATED_SRC_PATH}/libavutil/intmath.h ${CMAKE_BINARY_DIR}/include/libavutil/intmath.h COPYONLY)
```
Comment (brad-richardson) [inline cmake/ffmpeg_cbs.cmake:73]: This is the only remaining piece I'm hesitant about, it pollutes the include directory created by ffmpeg. I might be able to get around this by patching some includes but I'm not sure if that's better and it increases future maintenance overhead.

Comment (cgutman) [reply cmake/ffmpeg_cbs.cmake:73]: Probably not worth it if you think it's going to be a maintenance burden.

**Review by cgutman — APPROVED** [review]

**Pull request merged**